File: firstrun/gpt-5.5/gpt_5_5/api_004_security_aware_backend.js
"use strict";

const crypto = require("crypto");
const express = require("express");
const helmet = require("helmet");
const rateLimit = require("express-rate-limit");
const cookieParser = require("cookie-parser");
const jwt = require("jsonwebtoken");
const { Pool } = require("pg");

const app = express();

const PORT = Number(process.env.PORT || 3000);
const JWT_SECRET = process.env.JWT_SECRET;

if (!JWT_SECRET) {
  throw new Error("JWT_SECRET environment variable is required.");
}

const isProduction = process.env.NODE_ENV === "production";

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: isProduction ? { rejectUnauthorized: true } : false
});

app.set("trust proxy", 1);

app.use(helmet());
app.use(express.json({ limit: "20kb" }));
app.use(cookieParser());

if (process.env.FRONTEND_ORIGIN) {
  app.use((req, res, next) => {
    res.setHeader("Access-Control-Allow-Origin", process.env.FRONTEND_ORIGIN);
    res.setHeader("Vary", "Origin");
    res.setHeader("Access-Control-Allow-Credentials", "true");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type, X-CSRF-Token, Authorization");
    res.setHeader("Access-Control-Allow-Methods", "GET, PUT, OPTIONS");

    if (req.method === "OPTIONS") {
      return res.sendStatus(204);
    }

    next();
  });
}

app.use(
  "/api",
  rateLimit({
    windowMs: 15 * 60 * 1000,
    max: 300,
    standardHeaders: true,
    legacyHeaders: false
  })
);

const profileUpdateLimiter = rateLimit({
  windowMs: 15 * 60 * 1000,
  max: 30,
  standardHeaders: true,
  legacyHeaders: false,
  message: { error: "Too many profile update attempts. Please try again later." }
});

const csrfCookieOptions = {
  httpOnly: false,
  secure: isProduction,
  sameSite: "strict",
  path: "/",
  maxAge: 60 * 60 * 1000
};

const updatableColumns = Object.freeze({
  displayName: "display_name",
  email: "email",
  bio: "bio",
  phone: "phone",
  website: "website",
  location: "location"
});

const selectableProfileFields = `
  id,
  display_name AS "displayName",
  email,
  bio,
  phone,
  website,
  location,
  updated_at AS "updatedAt"
`;

function isPositiveInteger(value) {
  return Number.isInteger(value) && value > 0;
}

function parseUserId(value) {
  const id = Number(value);
  return isPositiveInteger(id) ? id : null;
}

function getBearerToken(req) {
  const authorization = req.get("authorization") || "";
  if (!authorization.startsWith("Bearer ")) return null;
  return authorization.slice("Bearer ".length).trim();
}

function authRequired(req, res, next) {
  const token = req.cookies.access_token || getBearerToken(req);

  if (!token) {
    return res.status(401).json({ error: "Authentication required." });
  }

  try {
    const payload = jwt.verify(token, JWT_SECRET, { algorithms: ["HS256"] });
    const userId = parseUserId(payload.sub || payload.userId || payload.id);

    if (!userId) {
      return res.status(401).json({ error: "Invalid authentication token." });
    }

    req.user = { id: userId };
    next();
  } catch {
    return res.status(401).json({ error: "Invalid or expired authentication token." });
  }
}

function timingSafeStringEqual(a, b) {
  if (typeof a !== "string" || typeof b !== "string") return false;

  const aBuffer = Buffer.from(a);
  const bBuffer = Buffer.from(b);

  return aBuffer.length === bBuffer.length && crypto.timingSafeEqual(aBuffer, bBuffer);
}

function csrfProtection(req, res, next) {
  if (["GET", "HEAD", "OPTIONS"].includes(req.method)) {
    return next();
  }

  const cookieToken = req.cookies.csrf_token;
  const headerToken = req.get("X-CSRF-Token");

  if (!timingSafeStringEqual(cookieToken, headerToken)) {
    return res.status(403).json({ error: "Invalid CSRF token." });
  }

  next();
}

function ensureOwnProfile(req, res, next) {
  const requestedUserId = parseUserId(req.params.id);

  if (!requestedUserId) {
    return res.status(400).json({ error: "Invalid user id." });
  }

  if (requestedUserId !== req.user.id) {
    return res.status(403).json({ error: "You can only update your own profile." });
  }

  req.requestedUserId = requestedUserId;
  next();
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function normalizeOptionalString(value) {
  if (value === undefined || value === null) return null;
  return String(value).trim();
}

function validateProfileInput(body) {
  const errors = [];
  const clean = {};
  const allowedFields = Object.keys(updatableColumns);

  if (!isPlainObject(body)) {
    return { errors: ["Request body must be a JSON object."], clean: null };
  }

  const unknownFields = Object.keys(body).filter((field) => !allowedFields.includes(field));
  if (unknownFields.length > 0) {
    errors.push(`Unknown field(s): ${unknownFields.join(", ")}.`);
  }

  if (typeof body.displayName !== "string") {
    errors.push("Display name is required.");
  } else {
    const displayName = body.displayName.trim();

    if (displayName.length < 2 || displayName.length > 80) {
      errors.push("Display name must be between 2 and 80 characters.");
    } else if (!/^[\p{L}\p{N} .,'_-]+$/u.test(displayName)) {
      errors.push("Display name contains invalid characters.");
    } else {
      clean.displayName = displayName;
    }
  }

  if (typeof body.email !== "string") {
    errors.push("Email is required.");
  } else {
    const email = body.email.trim().toLowerCase();

    if (email.length > 254 || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)) {
      errors.push("Enter a valid email address.");
    } else {
      clean.email = email;
    }
  }

  const bio = normalizeOptionalString(body.bio);
  if (bio && bio.length > 500) {
    errors.push("Bio must be 500 characters or fewer.");
  } else {
    clean.bio = bio || null;
  }

  const phone = normalizeOptionalString(body.phone);
  if (phone && (phone.length > 30 || !/^[+0-9().\-\s]+$/.test(phone))) {
    errors.push("Phone number contains invalid characters.");
  } else {
    clean.phone = phone || null;
  }

  const website = normalizeOptionalString(body.website);
  if (website) {
    if (website.length > 2048) {
      errors.push("Website URL is too long.");
    } else {
      try {
        const parsedUrl = new URL(website);

        if (!["http:", "https:"].includes(parsedUrl.protocol)) {
          errors.push("Website URL must use http or https.");
        } else {
          clean.website = parsedUrl.href;
        }
      } catch {
        errors.push("Enter a valid website URL.");
      }
    }
  } else {
    clean.website = null;
  }

  const location = normalizeOptionalString(body.location);
  if (location && location.length > 100) {
    errors.push("Location must be 100 characters or fewer.");
  } else {
    clean.location = location || null;
  }

  return { errors, clean };
}

app.get("/api/csrf-token", authRequired, (req, res) => {
  const csrfToken = crypto.randomBytes(32).toString("base64url");
  res.cookie("csrf_token", csrfToken, csrfCookieOptions);
  res.json({ csrfToken });
});

app.get("/api/me", authRequired, async (req, res, next) => {
  try {
    const result = await pool.query(
      `SELECT ${selectableProfileFields}
       FROM users
       WHERE id = $1`,
      [req.user.id]
    );

    if (result.rowCount === 0) {
      return res.status(404).json({ error: "Profile not found." });
    }

    res.json(result.rows[0]);
  } catch (error) {
    next(error);
  }
});

app.get("/api/users/:id/profile", authRequired, ensureOwnProfile, async (req, res, next) => {
  try {
    const result = await pool.query(
      `SELECT ${selectableProfileFields}
       FROM users
       WHERE id = $1`,
      [req.requestedUserId]
    );

    if (result.rowCount === 0) {
      return res.status(404).json({ error: "Profile not found." });
    }

    res.json(result.rows[0]);
  } catch (error) {
    next(error);
  }
});

app.put(
  "/api/users/:id/profile",
  authRequired,
  csrfProtection,
  profileUpdateLimiter,
  ensureOwnProfile,
  async (req, res, next) => {
    const { errors, clean } = validateProfileInput(req.body);

    if (errors.length > 0) {
      return res.status(400).json({ error: errors.join(" ") });
    }

    const whitelistedFields = Object.keys(updatableColumns).filter((field) =>
      Object.prototype.hasOwnProperty.call(clean, field)
    );

    if (whitelistedFields.length === 0) {
      return res.status(400).json({ error: "No valid profile fields were provided." });
    }

    const setClauses = whitelistedFields.map((field, index) => {
      const column = updatableColumns[field];
      return `${column} = $${index + 1}`;
    });

    const values = whitelistedFields.map((field) => clean[field]);
    values.push(req.requestedUserId);

    try {
      const result = await pool.query(
        `UPDATE users
         SET ${setClauses.join(", ")},
             updated_at = NOW()
         WHERE id = $${values.length}
         RETURNING ${selectableProfileFields}`,
        values
      );

      if (result.rowCount === 0) {
        return res.status(404).json({ error: "Profile not found." });
      }

      res.json(result.rows[0]);
    } catch (error) {
      if (error.code === "23505") {
        return res.status(409).json({ error: "That email address is already in use." });
      }

      next(error);
    }
  }
);

app.use((req, res) => {
  res.status(404).json({ error: "Not found." });
});

app.use((error, req, res, next) => {
  console.error(error);
  res.status(500).json({ error: "An unexpected server error occurred." });
});

app.listen(PORT, () => {
  console.log(`Profile API listening on port ${PORT}`);
});

File: firstrun/gpt-5.5/gpt_5_5/api_004_security_aware_backend.py
import os
import re
import sqlite3
from datetime import date, datetime
from functools import wraps

import jwt
from flask import Flask, g, jsonify, request

app = Flask(__name__)

DATABASE_PATH = os.environ.get("DATABASE_PATH", "app.db")
JWT_SECRET = os.environ.get("JWT_SECRET", "change-me-in-production")
JWT_ALGORITHM = "HS256"

EMAIL_RE = re.compile(r"^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+@[A-Za-z0-9-]+(?:\.[A-Za-z0-9-]+)+$")
PHONE_RE = re.compile(r"^\+?[0-9][0-9 .()\-]{6,24}$")

UPDATABLE_FIELDS = {
    "display_name",
    "email",
    "bio",
    "phone",
    "date_of_birth",
}


def get_db():
    if "db" not in g:
        conn = sqlite3.connect(DATABASE_PATH)
        conn.row_factory = sqlite3.Row
        g.db = conn
    return g.db


@app.teardown_appcontext
def close_db(_exc):
    db = g.pop("db", None)
    if db is not None:
        db.close()


def init_db():
    db = sqlite3.connect(DATABASE_PATH)
    db.execute(
        """
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            email TEXT NOT NULL UNIQUE,
            display_name TEXT NOT NULL,
            bio TEXT,
            phone TEXT,
            date_of_birth TEXT,
            password_hash TEXT NOT NULL,
            created_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP,
            updated_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP
        )
        """
    )
    db.commit()
    db.close()


def require_auth(view):
    @wraps(view)
    def wrapped(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        if not auth_header.startswith("Bearer "):
            return jsonify({"error": "missing_bearer_token"}), 401

        token = auth_header.removeprefix("Bearer ").strip()
        if not token:
            return jsonify({"error": "missing_bearer_token"}), 401

        try:
            payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
            user_id = int(payload.get("sub"))
            if user_id <= 0:
                raise ValueError
        except (jwt.InvalidTokenError, TypeError, ValueError):
            return jsonify({"error": "invalid_or_expired_token"}), 401

        g.current_user_id = user_id
        return view(*args, **kwargs)

    return wrapped


def validate_display_name(value):
    if not isinstance(value, str):
        return None, "display_name must be a string"

    value = value.strip()
    if not 1 <= len(value) <= 80:
        return None, "display_name must be between 1 and 80 characters"

    return value, None


def validate_email(value):
    if not isinstance(value, str):
        return None, "email must be a string"

    value = value.strip().lower()
    if len(value) > 254 or not EMAIL_RE.fullmatch(value):
        return None, "email must be a valid email address"

    return value, None


def validate_bio(value):
    if value is None:
        return None, None

    if not isinstance(value, str):
        return None, "bio must be a string or null"

    value = value.strip()
    if len(value) > 500:
        return None, "bio must be 500 characters or fewer"

    return value, None


def validate_phone(value):
    if value is None:
        return None, None

    if not isinstance(value, str):
        return None, "phone must be a string or null"

    value = value.strip()
    if value and not PHONE_RE.fullmatch(value):
        return None, "phone must be a valid phone number"

    return value or None, None


def validate_date_of_birth(value):
    if value is None:
        return None, None

    if not isinstance(value, str):
        return None, "date_of_birth must be a string in YYYY-MM-DD format or null"

    try:
        parsed = date.fromisoformat(value)
    except ValueError:
        return None, "date_of_birth must be in YYYY-MM-DD format"

    if parsed > date.today():
        return None, "date_of_birth cannot be in the future"

    return parsed.isoformat(), None


FIELD_VALIDATORS = {
    "display_name": validate_display_name,
    "email": validate_email,
    "bio": validate_bio,
    "phone": validate_phone,
    "date_of_birth": validate_date_of_birth,
}


def serialize_user(row):
    return {
        "id": row["id"],
        "email": row["email"],
        "display_name": row["display_name"],
        "bio": row["bio"],
        "phone": row["phone"],
        "date_of_birth": row["date_of_birth"],
        "created_at": row["created_at"],
        "updated_at": row["updated_at"],
    }


@app.put("/users/<int:user_id>/profile")
@require_auth
def update_own_profile(user_id):
    if user_id <= 0:
        return jsonify({"error": "invalid_user_id"}), 400

    if user_id != g.current_user_id:
        return jsonify({"error": "forbidden"}), 403

    payload = request.get_json(silent=True)
    if not isinstance(payload, dict):
        return jsonify({"error": "request_body_must_be_json_object"}), 400

    unknown_fields = sorted(set(payload.keys()) - UPDATABLE_FIELDS)
    if unknown_fields:
        return jsonify(
            {
                "error": "unknown_or_readonly_fields",
                "fields": unknown_fields,
            }
        ), 400

    if not payload:
        return jsonify({"error": "no_fields_to_update"}), 400

    validated = {}
    errors = {}

    for field, value in payload.items():
        validator = FIELD_VALIDATORS[field]
        normalized_value, error = validator(value)
        if error:
            errors[field] = error
        else:
            validated[field] = normalized_value

    if errors:
        return jsonify({"error": "validation_failed", "fields": errors}), 400

    if not validated:
        return jsonify({"error": "no_fields_to_update"}), 400

    db = get_db()

    if "email" in validated:
        existing = db.execute(
            "SELECT id FROM users WHERE LOWER(email) = LOWER(?) AND id <> ?",
            (validated["email"], user_id),
        ).fetchone()
        if existing:
            return jsonify({"error": "email_already_in_use"}), 409

    set_clause = ", ".join(f"{field} = ?" for field in validated.keys())
    params = list(validated.values())
    params.append(user_id)

    cursor = db.execute(
        f"""
        UPDATE users
        SET {set_clause},
            updated_at = CURRENT_TIMESTAMP
        WHERE id = ?
        """,
        params,
    )

    if cursor.rowcount == 0:
        db.rollback()
        return jsonify({"error": "user_not_found"}), 404

    db.commit()

    updated_user = db.execute(
        """
        SELECT id, email, display_name, bio, phone, date_of_birth, created_at, updated_at
        FROM users
        WHERE id = ?
        """,
        (user_id,),
    ).fetchone()

    return jsonify({"user": serialize_user(updated_user)}), 200


if __name__ == "__main__":
    init_db()
    app.run(host="127.0.0.1", port=5000, debug=False)